### ISSUE TITLE
Overloading between static and instance methods

### ISSUE THREAD

**Issue by mmomtchev** [swig/swig#2544]
Here is an extract of the `smart_pointer_static` test:
```cpp
class Foo2 { 
 public: 
  virtual ~Foo2() {}
  virtual int sum(int i, int j) { return i+j; }
  static int sum(int i, int j, int k); 
}; 
```

It includes an overloading with a static and and an instance method. This is something that used to be forbidden in C++03 but is now allowed. SWIG will consider this to be overloading and will generate an overloading dispatcher with two different cases.

Most languages however will have some very clear separation between the call signatures of the two methods - *which in fact could be identified even if the arguments were identical* as one must be called with the class, while the other one - with an object.

And at least in JS/V8 and in Python these will have to be registered in the class in a different manner - which means that they cannot share the same dispatcher.

Shouldn't these be simply two different methods?

**Comment by ojwb** [comment]
You can draw a straight line through any two points!

We really need to look at all the languages to get a clearer idea.  I checked a couple:

FWIW, in PHP you'd register as:

```
 PHP_ME(Foo2,sum,swig_arginfo_Foo2_sum,ZEND_ACC_PUBLIC)
```

vs

```
 PHP_ME(Foo2,sum,swig_arginfo_Foo2_sum,ZEND_ACC_PUBLIC | ZEND_ACC_STATIC)
```

Currently we register the overloaded form as the latter, and it doesn't work well for the non-`static` method:

```
$ php8.2 -n -d extension_dir=. -d extension=smart_pointer_static.so -d display_errors=stderr -r 'Foo2::sum(1,2,3);'
$ php8.2 -n -d extension_dir=. -d extension=smart_pointer_static.so -d display_errors=stderr -r 'Foo2::sum(1,2);'
Segmentation fault (core dumped)
$ php8.2 -n -d extension_dir=. -d extension=smart_pointer_static.so -d display_errors=stderr -r '$f=new Foo2(); $f->sum(1,2);'
Segmentation fault (core dumped)
$ php8.2 -n -d extension_dir=. -d extension=smart_pointer_static.so -d display_errors=stderr -r '$f=new Foo2(); $f::sum(1,2);'
Segmentation fault (core dumped)
```

I don't know if this could be made to work.  PHP rejected defining a static and non-static method in the same class in pure PHP, but sometimes the API allows things you can't do in pure PHP, or a clever dispatch function may be able to handle this.

Looks like Perl also registers once:

```
{"smart_pointer_staticc::Foo2_sum", _wrap_Foo2_sum},
```

And that magically works (looking at the code, I think its dispatch takes into account whether it gets a "this" pointer or not:

```
$ env LD_LIBRARY_PATH=.:$LD_LIBRARY_PATH  perl -I. -e 'use smart_pointer_static; smart_pointer_static::Foo2::sum(1,2,3)'
$ env LD_LIBRARY_PATH=.:$LD_LIBRARY_PATH  perl -I. -e 'use smart_pointer_static; smart_pointer_static::Foo2::sum(1,2)'
No matching function for overloaded 'Foo2_sum' at -e line 1.
$ env LD_LIBRARY_PATH=.:$LD_LIBRARY_PATH  perl -I. -e 'use smart_pointer_static; $f = smart_pointer_static::Foo2->new(); $f->sum(1,2,3)'
No matching function for overloaded 'Foo2_sum' at -e line 1.
$ env LD_LIBRARY_PATH=.:$LD_LIBRARY_PATH  perl -I. -e 'use smart_pointer_static; $f = smart_pointer_static::Foo2->new(); $f->sum(1,2)'
$
```

tl;dr: It seems some languages register static and non-static methods the same way so if this did get split, it seems it'd at least need to be optional per target language so that e.g. Perl could opt out.

**Comment by mmomtchev** [comment]
Python is more like Perl than JS in this case, but it will require modifying the dispatcher for it to work

**Comment by ojwb** [comment]
#1773 is probably the same underlying issue.

**Comment by mmomtchev** [comment]
I am still unsure what to do with the Python case, because for JS I am fixing something that either didn't compile at all or compiled but crashed, for Perl I made sure that it continued to work as it did before, for PHP I forbade something that compiled but crashed - however for Python I have to completely rework the two dispatchers to merge them in a single one that can handle both static and instance methods - all that for an edge case that no one noticed - and I wonder if the cure is not worse than the disease.

**Comment by ojwb** [comment]
> all that for an edge case that no one noticed

I'd agree it's an edge case, but #1773 and #796 suggest at least two people noticed, and for two different target languages (Javascript and Go).

While I'd certainly not complain if you fixed this for all target languages, I don't think you should feel obliged to - so long as the testsuite passes (e.g. by disabling part of the testcase for python) it seems reasonable to leave addressing this for Python to someone who's more familiar with that target language backend.

I think this probably could be made to work for PHP, by making the dispatch function a non-static method in PHP.  You could then only call the static method via a PHP object, but that's better than a crash.  I can take a look at that once the dust has settled here.

Incidentally:
> This is something that used to be forbidden in C++03

Do you have a reference for that?  Both GCC and clang seem happy to compile the code in your initial comment even with `-std=c++03` or `-std=c++98`.

**Comment by mmomtchev** [comment]
C++03 specifications:
https://www.lirmm.fr/~ducour/Doc-objets/ISO+IEC+14882-1998.pdf
Page 209, Section 13.1.2

The NAPI backend was the first one to not even compile - that's why I noticed it. This test didn't have a run check.

**Comment by wsfulton** [comment]
This is such an edge case and hard to support. I suggest closing as not supported and suggest workaround of using %ignore or %rename.

**Comment by ojwb** [comment]
@wsfulton Should `Examples/test-suite/smart_pointer_static.i` be reworked not to exercise it then?

**Comment by ojwb** [comment]
I had a quick look at the situation for SWIG/PHP - if the generated wrapper omits `ZEND_ACC_STATIC` for the dispatch method for this situation then you can call the non-static and static versions of the method successfully via a PHP object (like in C++, you can call a static method via an object in PHP).

You can't call the static version from PHP without an object like you could normally (i.e. without the overload) though.  So the crash is pretty easy to fix to give a reasonable behaviour, and I think that's worthwhile doing unless we're going to change to give a hard error in this situation (which I'm not entirely sold on doing given this currently works for some target languages, there's an instance of it in our testsuite and we've had at least two independent reports of it not working for different target languages: Javascript and Go).

**Comment by wsfulton** [comment]
The paragraph in the standard is saying that overloading static and non-static is ill-formed in certain circumstances, that is only when member function declarations have the same name and the same parameter types. gcc then warns (for newer standards too):
```
g++ -std=c++17 runme.cxx -o runme
runme.cxx:7:10: error: ‘void X::f()’ cannot be overloaded
     void f();
          ^
runme.cxx:6:17: error: with ‘static void X::f()’
     static void f();
                 ^
```

So the SWIG testcase is valid C++ as the static vs non-static methods have different parameters.

Given the varied nature of the target languages and difficulty in supporting this in some of them, let's just support it where we can and otherwise suggest %ignore/%rename, and test this. A full table of the languages that this works in would be useful to have in this issue.

**Comment by ojwb** [comment]
Oh I see, though probably the more nuanced point here is you can't overload e.g. `const X* X::f() const;` and `static X* f();` (you also can't overload e.g. `const X* X::f();` and `X* X::f();` but you can overload `const X* X::f() const;` and `X* X::f();` - essentially the overload there is resolved based on the type of the implicit `this` parameter).

> Given the varied nature of the target languages and difficulty in supporting this in some of them, let's just support it where we can and otherwise suggest %ignore/%rename, and test this. A full table of the languages that this works in would be useful to have in this issue.

This seems a sensible approach.

I'll take a look at fixing SWIG/PHP to handle this as it doesn't look hard and the current runtime crash ideally needs to be dealt with somehow.

**Comment by mmomtchev** [comment]
Ok, so what is your take then? Keep the current frontend behavior - that is generating a dispatcher for all methods that share the same name - and then have the language backend emit a warning/error when encountering this situation?

And remove this from `Examples/test-suite/smart_pointer_static.i` - where it doesn't seem to serve any purpose anyway?

**Comment by ojwb** [comment]
SWIG/PHP now reliably omits `ZEND_ACC_STATIC` for the dispatch wrapper function in this situation.  Looks like before it would based on whether the final function in the chain for that function was static or not.

> And remove this from Examples/test-suite/smart_pointer_static.i - where it doesn't seem to serve any purpose anyway?

I essentially asked that above, but wsfulton didn't answer.

I wrote a simple `smart_pointer_static_runme.php` as a regression test for my fix, but I can make a new testcase with such an overload if we want to remove it from `smart_pointer_static.i`.
